### ISSUE TITLE
Metadata allows generating invalid graphql schemas

### ISSUE THREAD

**Issue by sean-bennett112** [hasura/graphql-engine#6528]
I ran into this issue a bit while trying to use some other graphql tooling, specifically the [Apollo Devtools](https://github.com/apollographql/apollo-client-devtools). I was getting syntax errors whenever it would introspect the schema, and after pulling the schema down and generating the SDL, it turns out that the SDL was invalid.

For example, it included sections like the below examples:

```graphql
"""
update columns of table "channels"
"""
enum channels_update_column
```

```graphql
"""
input type for updating data in table "channels"
"""
input channels_set_input
```

which appear to be syntactically invalid(notes on that below).

It looks like the cause of this are metadata sections like the below:

```yaml
  update_permissions:
  - role: user
    permission:
      columns: []
      filter: {}
      check: null
```

While you'd think that just removing those sections would be enough to fix them, it looks like they're necessary in order to allow upsert semantics where we want to do nothing on conflict `(insert with on_conflict: {update_columns: []})`. NOTE: I'd be happy to be proven wrong about that point!

In any case, I think it makes sense to not allow metadata that results in an invalid graphql schema.

I'll note that, while I first discovered this in Apollo Devtools, I later tested it via both:

- `graphqurl`
- a combination of `get-graphql-schema` and `graphql-inspector`

both of which failed. `graphqurl` refused to print the schema out, giving:

```
Executing query... error
Error:  Error: channels_update_column values must be an object with value names as keys.
```

`get-graphql-schema` printed out the schema without complaint, but `graphql-inspector` failed flagging similar issues.

Both of them failed somewhere in `node_modules/graphql/`, which tells me that they're probably using the same core validation logic, and thus either that validation is wrong, or the introspection JSON from Hasura is invalid.

A further detail, just in case it's helpful at all, is that different tools produced slightly different schemas for the same introspection JSON. `get-graphql-schema` produced examples like the above, whereas the Apollo Devtools seems to produce blocks more like:

```graphql
"""
input type for updating data in table "channels"
"""
input channels_set_input {
}
```

I hope all that helps. I'm currently on 1.3.0, and haven't had a chance to test it on 1.3.3 yet.

**Comment by nicuveo** [comment]
This is a problem we have seen before: we didn't check whether some constructions could result in empty objects / enums / input objects. We fixed most of them during the rewrite that led to 2.0, so I guess technically this issue is a "won't fix" since it's about 1.3. Furthermore, post v2, we've fixed several bugs related to empty objects in the schema. However, I decided to have a quick look at the status of the code to see if there are things we can improve, to prevent bugs like this from arising in the future.

### Enums
For enums, our parser expects [a non-empty list of values](https://github.com/hasura/graphql-engine/blob/1243da1d54a0f122a68064668be566e04d790dfc/server/src-lib/Hasura/GraphQL/Parser/Internal/Input.hs#L242), meaning we simply cannot construct an empty enum in the schema anymore.

### Input objects
There, however, no such luck: `InputFieldsParser` stores [a list of definitions](https://github.com/hasura/graphql-engine/blob/1243da1d54a0f122a68064668be566e04d790dfc/server/src-lib/Hasura/GraphQL/Parser/Internal/Input.hs#L43); I don't think it's possible to change that and keep the applicative instance meaningful; we could, however change their [definition](https://github.com/hasura/graphql-engine/blob/1243da1d54a0f122a68064668be566e04d790dfc/server/src-lib/Hasura/GraphQL/Parser/Schema.hs#L441) to instead contain a non-empty list. Upside of this: we would fail to build schemas that would result in such invalid objects; downside: there'd be no way for users to still use Hasura with the invalid schema as they can today.

Another idea would be to change the `object` combinator (we even have [a TODO in the code for it](https://github.com/hasura/graphql-engine/blob/1243da1d54a0f122a68064668be566e04d790dfc/server/src-lib/Hasura/GraphQL/Parser/Internal/Input.hs#L267)): one way of doing this would be to return a `Maybe (Parser 'Input m a)`; each call site would then be forced to deal with the fact that it might not be representable. That seems fairly straightforward, I'll give it a try to see if there are any unforeseen issues there.

### Objects
Similarly, our [selection combinators](https://github.com/hasura/graphql-engine/blob/1243da1d54a0f122a68064668be566e04d790dfc/server/src-lib/Hasura/GraphQL/Parser/Internal/Parser.hs#L143) take a list, not a non-empty list. As the comment there suggests, a possibility would be to take non-empty lists everywhere, to enforce at every call site that we are creating objects that make sense.

We are gonna run into one known issue if we do this however: the empty `query_root` object. In short: the spec has an inherent contradiction; it both states that:
- the `query_root` object must always be present in the schema
- objects cannot be empty

As a result, it is impossible to build a "mutation only" server. We of course currently allow it, and generate an empty `query_root` object when that's the case. Most tools do actually have a workaround for that particular case, as it is quite common. However, if we do change our selection combinators, then we won't be able to express the query root at all. One potential solution would be to resurrect my internal pr that [added a placeholder in the case of an empty root](https://github.com/hasura/graphql-engine-mono/pull/148).

### tl;dr

- we won't solve this in 1.3, as most of those problems were fixed in 2.0
- enums are safe by construction
- I'll try to quickly see if we can make input objects safe by construction
- to make objects safe by construction, we would have to solve the problem of the empty root

**Comment by jkachmar** [comment]
@sean-bennett112 I'm closing this in light of @nicuveo's comment above, but please feel free to re-open it for discussion if you aren't satisfied with the resolution!
